wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,249 @@
+use std::sync::{Arc, RwLock};
+
+use crate::{
+    request::{
+        endpoint::ApiBaseUrl, RequestExecutor, RequestMethod, WpNetworkHeaderMap, WpNetworkRequest,
+        WpNetworkRequestBody, WpNetworkResponse,
+    },
+    RequestExecutionError, WpLoginCredentials,
+};
+
+#[async_trait::async_trait]
+pub trait Authenticator: Send + Sync + std::fmt::Debug {
+    // TODO: Use Result instead
+    async fn authenticate(&self, request: &mut WpNetworkRequest) -> bool;
```

Comment:
This sort of approach could work, but ideally these methods would return the necessary information rather than mutate a parameter. Even though Rust does a good job of limiting the issues with mutation, we should still prefer unit functions when it's possible to achieve the same result. There are various benefits to unit functions, but the one I am most interested in this case is avoiding the `Authenticator` trait being abused to make changes besides adding authentication headers.

In this case, I believe it'll be possible to return headers to be included in the request, rather than mutate the request. Let me know if I am missing anything.

_Note that I only had a very quick look at this since it's still Sunday for me. I can properly review tomorrow if necessary._

Reply:
I added a TODO here to remind me to change the mut reference to a Result. As I mentioned in https://github.com/Automattic/wordpress-rs/pull/327#discussion_r1789251419, I can make these changes after you think the current approach is acceptable. (Otherwise, I'm just wasting my time on changing um-mergeable code).